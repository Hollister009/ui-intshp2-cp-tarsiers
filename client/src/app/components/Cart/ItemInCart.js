import React, { Component } from 'react';
import { Notify } from 'react-redux-notify';
import NotifyService from '../../../utils/notify.service';
import { productType } from '../../types';

import styles from './Cart.module.scss';

class ItemInCart extends Component {
  static propTypes = { item: productType };

  static defaultProps = { item: null };

  constructor(props) {
    super(props);
    this.item = props.item;
    this.state = {
      chosenColor: this.item.chosenColor,
      chosenSize: this.item.chosenSize,
      chosenQuantity: this.item.chosenQuantity,
      total: this.item.price
    };
  }

  increment = () => {
    const { _id, price } = this.item;
    const { updateCartItem } = this.props;

    this.setState(
      state => ({
        chosenQuantity: state.chosenQuantity + 1,
        total: (state.chosenQuantity + 1) * price
      }),
      () => {
        const newItem = { ...this.item, ...this.state };

        updateCartItem({ _id, newItem });
      }
    );
  };

  decrement = () => {
    const { _id, price } = this.item;
    const { updateCartItem } = this.props;
    const { chosenQuantity } = this.state;

    if (chosenQuantity > 1) {
      this.setState(
        state => ({
          chosenQuantity: state.chosenQuantity - 1,
          total: (state.chosenQuantity - 1) * price
        }),
        () => {
          const newItem = { ...this.item, ...this.state };

          updateCartItem({ _id, newItem });
        }
      );
    }
  };

  toggleSizes = (e, size) => {
    e.preventDefault();
    const { _id } = this.item;
    const { updateCartItem } = this.props;

    this.setState({ chosenSize: size }, () => {
      const newItem = { ...this.item, ...this.state };

      updateCartItem({ _id, newItem });
    });
  };

  toggleColors = color => {
    const { _id } = this.item;
    const { updateCartItem } = this.props;

    this.setState({ chosenColor: color }, () => {
      const newItem = { ...this.item, ...this.state };

      updateCartItem({ _id, newItem });
    });
  };

  itemRemove = () => {
    const { removeFromCart, createNotification } = this.props;

    removeFromCart(this.item);
    createNotification(NotifyService.cartRemove);
  };

  renderSizes = () => {
    const { chosenSize } = this.state;

    return this.item.sizes.map((element, index, array) => {
      const active =
        element === chosenSize ? `${styles.highlightedSize}` : `${styles.size}`;

      return (
        <React.Fragment key={element}>
          <a
            href="/"
            className={active}
            onClick={e => this.toggleSizes(e, element)}
          >
            {element}
          </a>
          {index + 1 !== array.length ? <span>-</span> : null}
        </React.Fragment>
      );
    });
  };

  renderColors = () => {
    const { chosenColor } = this.state;

    return this.item.colors.map(color => {
      const style = { backgroundColor: `${color}` };

      return (
        <span
          key={color}
          role="button"
          tabIndex="0"
          className={color === chosenColor ? styles.active : styles.color}
          style={style}
          onClick={() => this.toggleColors(color)}
          onKeyDown={() => this.toggleColors(color)}
        >
          {color}
        </span>
      );
    });
  };

  render() {
    const { chosenQuantity, total } = this.state;
    const quantityButtons = (
      <div className={styles.select_quantity}>
        <button type="button" onClick={this.increment} data-type="increment">
          +
        </button>
        {chosenQuantity}
        <button type="button" onClick={this.decrement} data-type="decrement">
          -
        </button>
      </div>
    );

    return (
      <React.Fragment>
        <Notify position={NotifyService.position.topRight} />
        <div className={styles.cart_block}>
          <img
            src={this.item.src}
            alt={this.item.title}
            className={styles.item_image}
          />
          <h3 className={styles.item_title}>{this.item.title}</h3>
          <div className={styles.select_colors}>
            <div className={styles.scrollColor}>{this.renderColors()}</div>
          </div>
          <div className={styles.select_sizes}>
            <div className={styles.scrollSize}>{this.renderSizes()}</div>
          </div>
          <div>{this.item.price}</div>
          {quantityButtons}
          <button
            type="button"
            className={styles.item_remove}
            onClick={this.itemRemove}
            data-type="remove-btn"
          >
            <i className="far fa-trash-alt" />
          </button>
          <h3 className={styles.item_total}>{total}</h3>
        </div>
      </React.Fragment>
    );
  }
}

export default ItemInCart;
