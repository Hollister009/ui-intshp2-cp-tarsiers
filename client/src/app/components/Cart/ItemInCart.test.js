import React from 'react';
import { shallow } from 'enzyme';
import ItemInCart from './ItemInCart';

const products = require('../../../mocks/products.json');

describe('<ItemInCart />', () => {
  let props;

  beforeEach(() => {
    props = {
      item: products[2],
      setColor: jest.fn(),
      setSize: jest.fn(),
      setQuantityAndTotal: jest.fn(),
      removeFromCart: jest.fn(),
      createNotification: jest.fn()
    };
  });

  it('should match its snapshot', () => {
    const wrapper = shallow(<ItemInCart {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  xit('should return null when item is falsy', () => {
    const wrapper = shallow(<ItemInCart {...props} />);

    expect(wrapper.html()).toBe(null);
  });

  xit('should be able to increment', () => {
    const wrapper = shallow(<ItemInCart {...props} />);
    const chosenQuantity = wrapper.props;

    wrapper.find('button[data-type="increment"]').simulate('click');
    expect(chosenQuantity()).toBe(chosenQuantity() + 1);
  });

  xit('should be able to decrement', () => {
    const wrapper = shallow(<ItemInCart {...props} />);
    const chosenQuantity = wrapper.props;
    const decBtn = wrapper.find('button[data-type="decrement"]');

    decBtn.simulate('click');
    expect(chosenQuantity).toBe(1);
    decBtn.simulate('click');
    expect(chosenQuantity).toBe(1);
  });

  it('should be able to call toggleSizes', () => {
    const wrapper = shallow(<ItemInCart {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'toggleSizes');
    const size = wrapper.find('.size').at(0);
    const evt = { preventDefault() {}, target: { size, innerText: 'S' } };

    expect(wrapper.state().sizeClicked).toBe('');
    size.simulate('click', evt);

    expect(wrapper.state().sizeClicked).toBe('s');
    expect(spy).toHaveBeenCalled();
  });

  it('should be able to call toggleColors', () => {
    const wrapper = shallow(<ItemInCart {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'toggleColors');
    const color = wrapper.find('.color').at(0);
    const evt = {
      preventDefault() {},
      target: { color, innerText: '#171717' }
    };

    expect(wrapper.state().activeColor).toBe('');
    color.simulate('click', evt);

    expect(wrapper.state().activeColor).toBe('#171717');
    expect(spy).toHaveBeenCalled();
  });

  it('should call removeFromCart when inCart is true', () => {
    const { item } = props;
    const evt = { preventDefault() {} };
    const wrapper = shallow(<ItemInCart {...props} />);
    const btn = wrapper.find('button[data-type="remove-btn"]');

    wrapper.setProps({ inCart: true });
    wrapper.update();

    btn.simulate('click', evt);
    expect(props.removeFromCart).toHaveBeenCalledWith(item);
  });
});
