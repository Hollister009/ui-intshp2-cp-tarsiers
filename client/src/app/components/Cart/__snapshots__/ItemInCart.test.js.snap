// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemInCart /> should match its snapshot 1`] = `
<Fragment>
  <Connect(Notify)
    position="TopRight"
  />
  <div
    className="cart_block"
  >
    <img
      alt="Nike Sports Suit"
      className="item_image"
      src="https://res.cloudinary.com/so/image/upload/v1548246979/products/cba0707c66f3b196408f9bf6e92cc460--knt-nike.jpg"
    />
    <div
      className="item_title"
    >
      Nike Sports Suit
    </div>
    <div
      className="select_colors"
    >
      <span
        className="color"
        key="#171717"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        style={
          Object {
            "backgroundColor": "#171717",
          }
        }
        tabIndex="0"
      >
        #171717
      </span>
    </div>
    <div
      className="select_sizes"
    >
      <a
        className="size"
        href="/"
        onClick={[Function]}
      >
        s
      </a>
      <span>
        -
      </span>
      <a
        className="size"
        href="/"
        onClick={[Function]}
      >
        m
      </a>
      <span>
        -
      </span>
      <a
        className="size"
        href="/"
        onClick={[Function]}
      >
        l
      </a>
      <span>
        -
      </span>
      <a
        className="size"
        href="/"
        onClick={[Function]}
      >
        xl
      </a>
    </div>
    <div>
      800$
    </div>
    <div>
      <div
        className="select_quantity"
      >
        <button
          data-type="increment"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          data-type="decrement"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
    <button
      className="item_remove"
      data-type="remove-btn"
      onClick={[Function]}
      type="button"
    >
      <i
        className="far fa-trash-alt"
      />
    </button>
    <div
      className="item_total"
    >
      NaN$
    </div>
  </div>
</Fragment>
`;
